File: src/algorithms/arrandstring/Permutations.java
package algorithms.arrandstring; 

import java.util.*; 

public class Permutations {
    /* Function: Return all the no-duplicate permutations of an array. 
     * Solution: If the current value is same as the previous one, if the previous one is not used, 
     *     then the next pass will add this in it and create duplicates. 
     */ 
    
    public List<List<Integer>> permute(int[] num) {
        List<List<Integer>> res = new ArrayList<List<Integer>>(); 
        if (num == null || num.length == 0) {
            return res; 
        }
        
        Arrays.sort(num); 
        boolean[] used = new boolean[num.length]; 
        List<Integer> list = new ArrayList<Integer>(); 
        permHelper(num, used, list, res); 
        return res; 
    }
    
    private void permHelper(int[] num, boolean[] used, List<Integer> list, List<List<Integer>> res) {
        if (list.size() == num.length) {
            res.add(new ArrayList<Integer>(list)); 
            return; 
        }
        
        for (int i = 0; i < num.length; ++i) {
            if (used[i] || (i != 0 && num[i] == num[i - 1] && !used[i - 1])) {
                continue; 
            }
            
            used[i] = true; 
            list.add(num[i]); 
            permHelper(num, used, list, res); 
            list.remove(list.size() - 1); 
            used[i] = false; 
        }
    }
}

File: src/CPP/src/leetcode/Permutations.cpp
/*
 * [Source] https://leetcode.com/problems/permutations/
 * [Difficulty]: Medium
 * [Tag]: Backtracking
 */

#include <iostream>

using namespace std;

// [Solution]: First sort. Use a boolean vector to indicate if current value is used or not. If previous is not used and is same as current, skip.
// [Corner Case]: Duplicate
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/arrandstring/PermutationsUnique.java
*/

/* Java solution
public class Solution {
    public List<List<Integer>> permuteUnique(int[] num) {
        List<List<Integer>> result=new ArrayList<List<Integer>>();
		if(num==null || num.length<=0) return result;
		Arrays.sort(num);
		List<Integer> list=new ArrayList<Integer>();
		permute(num, num.length, list, result);
		return result;
    }
	
	private void permute(int[] array, int idx, List<Integer> prev, List<List<Integer>> result){
		if(idx==0){
			result.add(prev);
			return;
		}
		int i=0;
		List<Integer> firList=new ArrayList<Integer>(prev);
		firList.add(array[i]);
		int[] firCp=remove(array, idx, i);
		permute(firCp, idx-1, firList, result);
		while(i<idx-1){
			i++;
			if(array[i]!=array[i-1]){
				List<Integer> list=new ArrayList<Integer>(prev);
				list.add(array[i]);
				int[] cp=remove(array, idx, i);
				permute(cp, idx-1, list, result);
			}
		}
	}
	
	private int[] remove(int[] array, int idx, int del){
		if(idx==1) return array;
		int[] copy=new int[array.length];
		for(int i=0; i<del; i++){
			copy[i]=array[i];
		}
		for(int i=del; i<idx-1; i++){
			copy[i]=array[i+1];
		}
		return copy;
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
